From: Borislav Petkov <bp@suse.de>
Date: Sat, 16 Dec 2017 17:50:52 +0100
Subject: x86/spec: Add IBRS control functions
Patch-mainline: submitted on 2018/1/9
References: bsc#1068032

... into a separate compilation unit.

Carved out from a patch by Tim Chen <tim.c.chen@linux.intel.com>.

Signed-off-by: Borislav Petkov <bp@suse.de>
---
 arch/x86/include/asm/spec_ctrl.h |  3 +++
 arch/x86/kernel/cpu/Makefile     |  1 +
 arch/x86/kernel/cpu/spec_ctrl.c  | 22 ++++++++++++++++++++++
 3 files changed, 26 insertions(+)

diff --git a/arch/x86/include/asm/spec_ctrl.h b/arch/x86/include/asm/spec_ctrl.h
index 5e8c4124abed..cae607a2fb6c 100644
--- a/arch/x86/include/asm/spec_ctrl.h
+++ b/arch/x86/include/asm/spec_ctrl.h
@@ -50,5 +50,8 @@
 .Lend_\@:
 .endm
 
+#else /* __ASSEMBLY__ */
+void x86_enable_ibrs(void);
+void x86_disable_ibrs(void);
 #endif /* __ASSEMBLY__ */
 #endif /* _ASM_X86_SPEC_CTRL_H */
diff --git a/arch/x86/kernel/cpu/Makefile b/arch/x86/kernel/cpu/Makefile
index 58031303e304..c8205ff2a0f3 100644
--- a/arch/x86/kernel/cpu/Makefile
+++ b/arch/x86/kernel/cpu/Makefile
@@ -16,6 +16,7 @@ obj-y			:= intel_cacheinfo.o scattered.o topology.o
 obj-y			+= common.o
 obj-y			+= rdrand.o
 obj-y			+= match.o
+obj-y			+= spec_ctrl.o
 
 obj-$(CONFIG_PROC_FS)	+= proc.o
 obj-$(CONFIG_X86_FEATURE_NAMES) += capflags.o powerflags.o
diff --git a/arch/x86/kernel/cpu/spec_ctrl.c b/arch/x86/kernel/cpu/spec_ctrl.c
new file mode 100644
index 000000000000..10c2c6fff3e5
--- /dev/null
+++ b/arch/x86/kernel/cpu/spec_ctrl.c
@@ -0,0 +1,22 @@
+/*
+ * Speculation control stuff
+ *
+ */
+
+#include <asm/msr.h>
+#include <asm/processor.h>
+#include <asm/spec_ctrl.h>
+
+void x86_disable_ibrs(void)
+{
+	if (boot_cpu_has(X86_FEATURE_SPEC_CTRL))
+		native_wrmsrl(MSR_IA32_SPEC_CTRL, 0);
+}
+EXPORT_SYMBOL_GPL(x86_disable_ibrs);
+
+void x86_enable_ibrs(void)
+{
+	if (boot_cpu_has(X86_FEATURE_SPEC_CTRL))
+		native_wrmsrl(MSR_IA32_SPEC_CTRL, FEATURE_ENABLE_IBRS);
+}
+EXPORT_SYMBOL_GPL(x86_enable_ibrs);

